### PR TITLE
Debian package continuation

### PR DESCRIPTION
**Pull request by ArneBab**

### CODE REVIEW

**Comment by desyncr** [comment]
TODO
- [ ] Generate seednodes.fref from https://github.com/freenet/seedrefs
- [ ] Move wrapper.conf somewhere else (but probably preserve in this repository)

**Comment by ArneBab** [comment]
@desyncr what’s the state of the debian package?

**Comment by desyncr** [comment]
@ArneBab Hey there Arne, yes the TODO points are up to date and I'll keep them up do date with each iteration. I'll continue working tackling down the low hanging fruits (missing attribution, instructions) then test and iterate over multiple OSes.
Hopefully getting to a point of testing soon (few weeks).

**Comment by ArneBab** [comment]
@desyncr thank you! :heart:

**Comment by desyncr** [comment]
Lintian warnings and errors pending:
```
Now running lintian freenet_0.7.5+1493-1_amd64.changes ...
E: freenet changes: bad-distribution-in-changes-file unstable
E: freenet: init.d-script-does-not-implement-required-option etc/init.d/freenet force-reload
E: freenet: init.d-script-does-not-implement-required-option etc/init.d/freenet restart
E: freenet: init.d-script-does-not-implement-required-option etc/init.d/freenet start
E: freenet: init.d-script-does-not-implement-required-option etc/init.d/freenet stop
E: freenet: init.d-script-missing-dependency-on-local_fs etc/init.d/freenet: required-stop
E: freenet: init.d-script-should-always-start-service etc/default/freenet (line 10)
E: freenet source: temporary-debhelper-file freenet.debhelper.log
W: freenet source: custom-compression-in-debian-rules dh_builddeb -- -Zgzip (line 46)
W: freenet: init.d-script-does-not-source-init-functions etc/init.d/freenet
W: freenet: initial-upload-closes-no-bugs
W: freenet: missing-systemd-service-for-init.d-script freenet
W: freenet source: package-needs-versioned-debhelper-build-depends 11
W: freenet: unknown-field freenet_0.7.5+1493-1_amd64.deb Tag
N: 26 hints overridden (26 errors)
Finished running lintian.
```

**Comment by ArneBab** [comment]
Nice progress!

**Comment by desyncr** [comment]
Packaging guide: https://github.com/desyncr/fred/wiki/Debian-package

**Comment by ArneBab** [comment]
@desyncr does this mean, you’re finished now?

**Comment by desyncr** [comment]
They're some errors and warnings still. But the package builds and works.

I'd say the package is ready to go since the errors are related to how the init script works with the wrapper.
Besides that since we don't intent to be in the debian's repositories I'd not care too much about these errors.

I'd like to have some feedback from users installing the"beta" versions.

**Comment by ArneBab** [comment]
@desyncr did you already get enough feedback or should I try it in a VM?

**Comment by desyncr** [comment]
@ArneBab I did some feedback regarding packagecloud.io not having the packages for different versions, which is true. I need to build a package and upload to the different debian and ubuntu versions.

My concern with packagecloud.io is that the disk quota will be exhausted very quickly if I have to upload the _same_ package for each version we want to support. Alternatives are very little as far as I'm concern:
1) Hosting it under the Freenet Project Inc umbrella will introduce monthly costs which I'd not like to incur.
2) Submitting to Debian will definitely take a long while since I'm not able to progress much on the linting issues, let alone any feedback the Debian packaging team might have.

Regardless of the linting issues, the package does build and runs correctly. I'm using it on my node without problems (although my node doesn't have a high uptime).

**Comment by ArneBab** [comment]
Can you share the current lint output? Maybe I can help get that fixed.

**Comment by desyncr** [comment]
@ArneBab Here they are, not much progress :(
```
Now running lintian freenet_0.7.5+1494-4_amd64.changes ...
E: freenet changes: bad-distribution-in-changes-file unstable
E: freenet: init.d-script-does-not-implement-required-option etc/init.d/freenet force-reload
E: freenet: init.d-script-does-not-implement-required-option etc/init.d/freenet restart
E: freenet: init.d-script-does-not-implement-required-option etc/init.d/freenet start
E: freenet: init.d-script-does-not-implement-required-option etc/init.d/freenet stop
E: freenet: init.d-script-missing-dependency-on-local_fs etc/init.d/freenet: required-stop
E: freenet: init.d-script-should-always-start-service etc/default/freenet (line 10)
W: freenet source: custom-compression-in-debian-rules dh_builddeb -- -Zgzip (line 42)
W: freenet: init.d-script-does-not-source-init-functions [etc/init.d/freenet]
W: freenet: missing-systemd-service-for-init.d-script freenet [etc/init.d/freenet]
W: freenet source: no-debian-changes
W: freenet source: no-versioned-debhelper-prerequisite 11
W: freenet: unknown-field Tag
N: 6 hints overridden (6 errors); 0 unused override
````

**Comment by desyncr** [comment]
Updated errors with some overrides:
```
E: freenet changes: bad-distribution-in-changes-file unstable
E: freenet: init.d-script-missing-dependency-on-local_fs etc/init.d/freenet: required-stop
E: freenet: init.d-script-should-always-start-service etc/default/freenet (line 10)
W: freenet source: debian-rules-should-not-use-custom-compression-settings dh_builddeb -- -Zgzip (line 42)
W: freenet: init.d-script-does-not-source-init-functions etc/init.d/freenet
W: freenet source: latest-debian-changelog-entry-reuses-existing-version 0.7.5+1494-7 ~= 0.7.5+1494-7 (last used: Sat, 17 Sep 2022 13:57:51 +0100)
W: freenet: missing-systemd-service-for-init.d-script freenet
W: freenet source: newer-standards-version 4.5.1 (current is 4.5.0)
W: freenet source: package-needs-versioned-debhelper-build-depends 11
N: 30 tags overridden (30 errors)
Finished running lintian.
```

Some notes on the error messages:
# E: freenet changes: bad-distribution-in-changes-file unstable

You&rsquo;ve specified an unknown target distribution for your upload in the debian/changelog file. It is possible that you are uploading for a different distribution than the one Lintian is checking for. In that case, passing &#x2013;profile $VENDOR may fix this warning.
([source](https://lintian.debian.org/tags/bad-distribution-in-changes-file))

In a .changes file or parsed changelog output this contains the (space-separated) name(s) of the distribution(s) where this version of the package should be installed. Valid distributions are determined by the archive maintainers. 10 The Debian archive software only supports listing a single distribution. Migration of packages to other distributions is handled outside of the upload process.
([source](https://www.debian.org/doc/debian-policy/ch-controlfields.html#distribution))

See [here](https://bugs.launchpad.net/ubuntu/+source/lintian/+bug/1303603/comments/7) for explanation on this behaviour.


# E: freenet: init.d-script-does-not-implement-required-option etc/init.d/freenet force-reload

The /etc/init.d scripts have to support the following command line arguments: start, stop, restart, force-reload.
([source](https://lintian.debian.org/tags/init.d-script-does-not-implement-required-option))

Checkout freenet/debian [example](https://github.com/freenet/debian/blob/master/debian/freenet-daemon.lintian-overrides):

    # lintian not smart enough to detect these things for service-wrapper-java
    freenet-daemon binary: init.d-script-does-not-source-init-functions etc/init.d/freenet-daemon
    freenet-daemon binary: init.d-script-does-not-implement-required-option etc/init.d/freenet-daemon start
    freenet-daemon binary: init.d-script-does-not-implement-required-option etc/init.d/freenet-daemon stop
    freenet-daemon binary: init.d-script-does-not-implement-required-option etc/init.d/freenet-daemon restart
    freenet-daemon binary: init.d-script-does-not-implement-required-option etc/init.d/freenet-daemon force-reload
    freenet-daemon binary: init.d-script-does-not-implement-optional-option etc/init.d/freenet-daemon status

    E: freenet: init.d-script-does-not-implement-required-option etc/init.d/freenet force-reload
    E: freenet: init.d-script-does-not-implement-required-option etc/init.d/freenet restart
    E: freenet: init.d-script-does-not-implement-required-option etc/init.d/freenet start
    E: freenet: init.d-script-does-not-implement-required-option etc/init.d/freenet stop

From [man](https://www.huge-man-linux.net/man1/lintian.html):

    For binary packages, Lintian looks for overrides in a file named
    usr/share/lintian/overrides/<package> inside the binary package, where
    <package> is the name of the binary package.  For source packages,
    Lintian looks for overrides in debian/source/lintian-overrides and then
    in debian/source.lintian-overrides if the first file is not found.  The
    first path is preferred.  See the Lintian User's Manual for the syntax
    of overrides.

**Comment by desyncr** [comment]
@ArneBab By the way, the build process can be seen [here](https://github.com/desyncr/fred/actions).

**Comment by ArneBab** [comment]
The Zgzip error sounds like it could be caused by this line:
```
override_dh_builddeb:
	dh_builddeb -- -Zgzip
```

**Comment by ArneBab** [comment]
@desyncr will you take care of the Zgzip error? What’s still missing to get this done?

File: build.gradle [hunk 1]
```diff
@@ -94,7 +94,7 @@ task jar (type: Jar, overwrite: true) {
         exclude 'freenet/node/Version.class'
         exclude 'freenet/node/Version$1.class'
     }
-    preserveFileTimestamps = false
+    preserveFileTimestamps = true
```
Comment (Thynix) [inline build.gradle:97]: What's the reasoning for setting this to `true`? For deterministic builds it needs to be `false` - [documentation of that is here](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.bundling.Zip.html#org.gradle.api.tasks.bundling.Zip:preserveFileTimestamps). If there's a need that it not be deterministic, I think we'd need another `task` to build for the Debian package.

**Review by Thynix — CHANGES_REQUESTED** [review]

**Comment by ArneBab** [comment]
@desyncr are you still working on this?

**Comment by ArneBab** [comment]
I looked into this again and I see no dangerous stuff, so I decided to merge it. I’ll undo the preserveFileTimestamps change on next.

**Pull request merged**